### PR TITLE
Support netty4 and5

### PR DESCRIPTION
**Pull request by louiscryan**
@nmittler A little hacky but this is the only change needs to work with both Netty 4 & 5

### CODE REVIEW

**Comment by nmittler** [comment]
ick! .... LGTM :)

**Comment by buchgr** [comment]
hmm ... I think this could be done nicer with reflections?

**Comment by buchgr** [comment]
Actually, this should be enough to work with both?

```java
    @Override
    public EventLoopGroup create() {
      Executor executor = Executors.newFixedThreadPool(nEventLoops, new ThreadFactoryBuilder().setNameFormat(name + "-%d").build());
      return new NioEventLoopGroup(nEventLoops, executor);
    }
```

**Comment by louiscryan** [comment]
Definitely better with reflection but we can't quite be as simple as Jakob's suggestion as Netty4 doesn't have a constructor that takes an executor.

**Comment by louiscryan** [comment]
PTAL

**Comment by buchgr** [comment]
Netty 4.1's NioEventLoopGroup does take an Executor but not an ExecutorServiceFactory, so this should work? https://github.com/netty/netty/blob/4.1/transport/src/main/java/io/netty/channel/nio/NioEventLoopGroup.java#L57

File: netty/src/main/java/io/grpc/transport/netty/Utils.java [hunk 1]
```diff
@@ -173,8 +177,22 @@ private static Http2Headers convertMetadata(Metadata headers) {
...
-          .build());
+      ThreadFactory threadFactory = new ThreadFactoryBuilder().setNameFormat(name + "-%d").build();
+      try {
+        Constructor<NioEventLoopGroup> constructor = NioEventLoopGroup.class.getConstructor(
+            Integer.TYPE, ThreadFactory.class);
+        if (constructor == null) {
+          // is netty 5
+          constructor = NioEventLoopGroup.class.getConstructor(Integer.TYPE, Executor.class);
+          return constructor.newInstance(nEventLoops,
+              Executors.newFixedThreadPool(nEventLoops, threadFactory));
```
Comment (buchgr) [inline netty/src/main/java/io/grpc/transport/netty/Utils.java]: @louiscryan if we pass in the `Executor` to the `EventLoopGroup`, we should also shut it down in close().

**Comment by louiscryan** [comment]
Good point @buchgr, not sure how I missed that. PTAL

File: netty/src/main/java/io/grpc/transport/netty/Utils.java (hunk 1 above)
Comment (louiscryan) [inline netty/src/main/java/io/grpc/transport/netty/Utils.java]: Is fixed in my more recent update. PTAL

File: netty/src/main/java/io/grpc/transport/netty/Utils.java [hunk 2]
```diff
@@ -173,8 +178,19 @@ private static Http2Headers convertMetadata(Metadata headers) {
 
     @Override
     public EventLoopGroup create() {
-      return new NioEventLoopGroup(nEventLoops, new ThreadFactoryBuilder().setNameFormat(name + "-%d")
-          .build());
+      // Use the executor based constructor so we can work with both Netty4 & Netty5.
+      ThreadFactory threadFactory = new ThreadFactoryBuilder().setNameFormat(name + "-%d").build();
+      int parallelism = nEventLoops == 0 ?
```
Comment (buchgr) [inline netty/src/main/java/io/grpc/transport/netty/Utils.java]: @louiscryan we don't that line anymore, the nEventLoops == 0 case is now part of the offical Netty API [1]. It's in 5 already, still have to submit a PR for 4.0 and 4.1, but will do today or monday.

https://github.com/netty/netty/commit/93bcc6bbdef53e9249802e1bba72aca06f90ad45

**Comment by buchgr** [comment]
@louiscryan LGTM

**Comment by nmittler** [comment]
Cherrypicked into master.

**Pull request closed without merging**